# Clean and analyze the price and mileage of German used cars listed in eBay

Datasets are usually not ready for analysis. They need to be cleaned first. In this notebook, a car listing from eBay Germany is analyzed. The analysis is focused on the price and mileages of the car brands listed in this website. What akward stuff the users fill in when posting a car ad? What are the most expensive or cheapests second-hand cars you can find there? Which second-hand is normally listed with a high or low mileage?

## Import libraries

In [1]:
import numpy as np
import pandas as pd

## Explore the data

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

### Look for missing (null) values

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

### Rename the columns with a more intuitive and clearer names

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Look for outliers and "weird" values

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In this analysis you can note the next:

- nr_of_pictures is all zero
- powerPS has inconsistent outliers
- registration_year max is 9999
- price and odomometer are string and must be numeric
- there are 5 columns with NaN values, sometimes up to ~20% of the data
- There are one row that it is not-repeated in seller and offer_type

### Convert the price and mileage in integers to make calculations afterwards

In [10]:
print(autos["price"].head())
print(autos["odometer"].head())

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object


In [11]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos.rename({"price":"price_usd"}, inplace=True, axis=1)

In [12]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
autos.rename({"odometer":"odometer_km"}, inplace=True, axis=1)

In [13]:
autos["odometer_km"].unique().shape

(13,)

In [14]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [15]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [16]:
autos["price_usd"].unique().shape

(2357,)

In [17]:
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In the cells below you can see that some cars are listed at $0 USD or even prices like $12345678 USD, which intuitevly don't make sense.

In [18]:
autos["price_usd"].value_counts().sort_index(ascending=True).head(100)

0      1421
1       156
2         3
3         1
5         2
       ... 
300     384
310       1
320      12
325       5
329       2
Name: price_usd, Length: 100, dtype: int64

In [19]:
autos["price_usd"].value_counts().sort_index(ascending=True).tail(100)

65699       1
65700       1
65990       1
66500       1
66964       1
           ..
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price_usd, Length: 100, dtype: int64

Let's put some boundaries for the prices, if the price is out of the boundaries, then make it a missing value (nan)

In [20]:
autos[~autos["price_usd"].between(49,350001)] = np.nan

In [21]:
autos["price_usd"].describe()

count     48351.000000
mean       5914.970776
std        9071.407458
min          49.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price_usd, dtype: float64

### When were more ads created, crawled or last seen?

In [22]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [23]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.02452
2016-03-06    0.01360
2016-03-07    0.03484
2016-03-08    0.03216
2016-03-09    0.03196
2016-03-10    0.03116
2016-03-11    0.03150
2016-03-12    0.03576
2016-03-13    0.01514
2016-03-14    0.03546
2016-03-15    0.03316
2016-03-16    0.02852
2016-03-17    0.03054
2016-03-18    0.01248
2016-03-19    0.03362
2016-03-20    0.03654
2016-03-21    0.03606
2016-03-22    0.03186
2016-03-23    0.03120
2016-03-24    0.02846
2016-03-25    0.03046
2016-03-26    0.03122
2016-03-27    0.03006
2016-03-28    0.03374
2016-03-29    0.03300
2016-03-30    0.03260
2016-03-31    0.03078
2016-04-01    0.03262
2016-04-02    0.03436
2016-04-03    0.03730
2016-04-04    0.03532
2016-04-05    0.01262
2016-04-06    0.00306
2016-04-07    0.00134
NaN           0.03298
Name: date_crawled, dtype: float64

In [24]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-04    0.03568
2016-04-05    0.01138
2016-04-06    0.00314
2016-04-07    0.00120
NaN           0.03298
Name: ad_created, Length: 77, dtype: float64

In [25]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.00104
2016-03-06    0.00418
2016-03-07    0.00524
2016-03-08    0.00710
2016-03-09    0.00932
2016-03-10    0.01026
2016-03-11    0.01198
2016-03-12    0.02296
2016-03-13    0.00858
2016-03-14    0.01222
2016-03-15    0.01536
2016-03-16    0.01592
2016-03-17    0.02716
2016-03-18    0.00708
2016-03-19    0.01528
2016-03-20    0.01994
2016-03-21    0.01992
2016-03-22    0.02070
2016-03-23    0.01798
2016-03-24    0.01912
2016-03-25    0.01854
2016-03-26    0.01622
2016-03-27    0.01506
2016-03-28    0.02016
2016-03-29    0.02158
2016-03-30    0.02392
2016-03-31    0.02304
2016-04-01    0.02212
2016-04-02    0.02406
2016-04-03    0.02434
2016-04-04    0.02370
2016-04-05    0.12086
2016-04-06    0.21434
2016-04-07    0.12774
NaN           0.03298
Name: last_seen, dtype: float64

### Cleaning the registration years

Come cars are listed for years that are even before the invention of the cars! And other are posted as if the cars were from the future. Let's correct this by putting some boundaries here too.

In [26]:
autos[~autos["registration_year"].between(1900,2016)] = np.nan

In [27]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1910.0    0.000043
1927.0    0.000022
1929.0    0.000022
1931.0    0.000022
1934.0    0.000043
            ...   
2012.0    0.028165
2013.0    0.017235
2014.0    0.014244
2015.0    0.008262
2016.0    0.025970
Name: registration_year, Length: 78, dtype: float64

## Explore the brands

In the cells below, there is a list of the car brands founded in this dataset, and finally, the sorted average price and mileages of those brands.

In [28]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', nan, 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [29]:
all_brands = autos["brand"].unique()

autos["brand"].value_counts()

volkswagen        9824
bmw               5110
opel              4987
mercedes_benz     4485
audi              4025
ford              3252
renault           2193
peugeot           1390
fiat              1195
seat               848
skoda              763
nissan             712
mazda              709
smart              659
citroen            653
toyota             593
hyundai            467
sonstige_autos     447
volvo              423
mini               408
mitsubishi         380
honda              365
kia                330
alfa_romeo         309
porsche            281
suzuki             276
chevrolet          265
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru              98
land_rover          98
saab                77
jaguar              73
daewoo              69
trabant             64
rover               62
lancia              49
lada                27
Name: brand, dtype: int64

In [30]:
price_per_brand = {}

for b in all_brands:
    selected_rows = autos[autos["brand"] == b]
    mean_price = selected_rows["price_usd"].mean()
    price_per_brand[b] = mean_price
    
{k: v for k, v in sorted(price_per_brand.items(), key=lambda item: item[1])}

{'daewoo': 1064.0579710144928,
 'rover': 1602.2903225806451,
 'daihatsu': 1636.1965811965813,
 'trabant': 1818.609375,
 'renault': 2483.889192886457,
 'lada': 2688.296296296296,
 'fiat': 2818.456066945607,
 'opel': 2996.0850210547424,
 'peugeot': 3100.6928057553955,
 'saab': 3211.6493506493507,
 'mitsubishi': 3430.25,
 'lancia': 3444.877551020408,
 'chrysler': 3465.743902439024,
 'smart': 3591.0212443095597,
 'ford': 3762.1165436654364,
 'citroen': 3784.924961715161,
 'subaru': 4033.7551020408164,
 'alfa_romeo': 4100.915857605178,
 'suzuki': 4111.572463768116,
 'mazda': 4112.596614950635,
 'honda': 4119.109589041096,
 'seat': 4423.116745283019,
 'nissan': 4750.063202247191,
 'volvo': 4993.208037825059,
 'toyota': 5167.091062394604,
 'hyundai': 5376.740899357602,
 'volkswagen': 5423.276872964169,
 'bmw': 8376.805088062622,
 nan: nan,
 'dacia': 5915.528455284553,
 'kia': 5982.330303030303,
 'skoda': 6393.018348623853,
 'chevrolet': 6709.358490566037,
 'mercedes_benz': 8663.059308807135,


In [31]:
mileage_per_brand = {}

for b in all_brands:
    selected_rows = autos[autos["brand"] == b]
    mean_mileage = selected_rows["odometer_km"].mean()
    mileage_per_brand[b] = mean_mileage
    
{k: v for k, v in sorted(mileage_per_brand.items(), key=lambda item: item[1])}

{'trabant': 55312.5,
 'lada': 83518.51851851853,
 'dacia': 84268.29268292683,
 'mini': 88308.82352941176,
 'sonstige_autos': 90313.19910514541,
 'porsche': 97811.38790035587,
 'chevrolet': 99547.16981132075,
 'smart': 99590.28831562975,
 'hyundai': 106541.75588865097,
 'suzuki': 108315.21739130435,
 'skoda': 110884.66579292268,
 'kia': 112530.30303030302,
 'toyota': 115944.35075885328,
 'daihatsu': 116410.2564102564,
 'fiat': 117066.94560669456,
 'land_rover': 118010.20408163265,
 'nissan': 118370.78651685393,
 'citroen': 119647.77947932619,
 'daewoo': 121231.88405797101,
 'seat': 121303.06603773584,
 'lancia': 122346.9387755102,
 'honda': 122493.1506849315,
 'jaguar': 124178.08219178082,
 'ford': 124295.81795817958,
 'mazda': 124464.03385049365,
 'subaru': 125612.24489795919,
 'peugeot': 127122.30215827338,
 'jeep': 127122.64150943396,
 'mitsubishi': 127157.8947368421,
 'renault': 128144.09484724123,
 'volkswagen': 128757.63436482084,
 'audi': 129254.65838509316,
 'opel': 129440.54541